test/framework/src/main/java/org/elasticsearch/plugins/MockPluginsService.java

```diff
@@ -76,4 +80,47 @@ public PluginsAndModules info() {
             List.of()
         );
     }
+
+    @Override
+    public <T> List<? extends T> loadServiceProviders(Class<T> service) {
+        // We use a set here to avoid duplicates because SPIClassIterator will match
+        // all plugins in MockNode, because all plugins are loaded by the same class loader.
+        Set<T> result = new HashSet<>();
+
+        for (LoadedPlugin pluginTuple : plugins()) {
+            result.addAll(createExtensions(service, pluginTuple.instance()));
+        }
+
+        return result.stream().toList();
+    }
+
+    /**
+     * When we load tests with MockNode, all plugins are loaded with the same class loader,
+     * which breaks loading service providers with our SPIClassIterator. Since all plugins are
+     * loaded in the same class loader, we find all plugins for any class found by the SPIClassIterator
+     * causing us to pass wrong plugin type to createExtension. This modified createExtensions, checks for
+     * the type and returns an empty list if the plugin class type is incompatible.
+     */
+    static <T> List<? extends T> createExtensions(Class<T> extensionPointType, Plugin plugin) {
+        SPIClassIterator<T> classIterator = SPIClassIterator.get(extensionPointType, plugin.getClass().getClassLoader());
+        List<T> extensions = new ArrayList<>();
+        while (classIterator.hasNext()) {
+            Class<? extends T> extensionClass = classIterator.next();
+
+            @SuppressWarnings("unchecked")
+            Constructor<T>[] constructors = (Constructor<T>[]) extensionClass.getConstructors();
+            boolean compatible = true;
```

Comment:
Why this leniency?

Reply:
When there's no constructor argument we simply fall though to load the plugin, there's no need to distinguish from which Plugin the load happened. We ensure we de-dup in the caller method by using a set, because the service provider will get loaded for every plugin.

I'll add a comment to this effect.